enhancements/olm/opm-mirroring.md

```diff
@@ -0,0 +1,183 @@
+
+
+### Goals
+* Define a mechanism for operators to specify related operand images that should be collected for disconnected installation.
+* Define the changes required in `olm` and `operator-registry` to allow those images to be discovered
+* Define `opm` commands which can build and mirror all images required for mirroring to a disconnected envorinment
+
+### Non-Goals
+* Provide an api for fine-grained control over which operators / operands are mirrored. 
+
+## Proposal
+
+The basic approach will be to:
+
+- Allow operator manifests to specify [related images](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/doc/contributors/design-proposals/related-images.md) in their manifests
+- Generate catalog container images that can replace the appregistry repositories
+- For each operator within the catalog image, mirror it to the disconnected environment
+
+
+### Specify related images
+
+We add a new field to the `ClusterServiceVersion` spec. 
+
+```yaml
+kind: ClusterServiceVersion 
+spec:
+  relatedImages:
+  - name: default
+    image: quay.io/coreos/etcd@sha256:12345 
+  - name: etcd-2.1.5
+    image: quay.io/coreos/etcd@sha256:12345 
+  - name: etcd-3.1.1
+    image: quay.io/coreos/etcd@sha256:12345  
+```
+
+This is covered in depth in the [related images](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/doc/contributors/design-proposals/related-images.md) proposal for OLM.
+
+### Expose related images
+
+`operator-registry` has tools for building an index of operator manifests into a sqlite artifact, with interfaces provided for querying that database.
+
+A new table will be added to the sql schema:
+
+```sql
+	CREATE TABLE IF NOT EXISTS related_image (
+	    image TEXT,
+		operatorbundle_name TEXT,
+	    FOREIGN KEY(operatorbundle_name) REFERENCES operatorbundle(name)
+	);
+```
+
+Two new queries should be added to allow effecient collection of related images:
+
+```go
+type Query interface { 
+  ListImages(ctx context.Context) ([]string, error)
+  GetImagesForBundle(ctx context.Context, operatorName string) ([]string, error)
+  // the rest of the existing interface
+}
+```
+
+The loading process will read the new fields in the `ClusterServiceVersion` and populate the `bundle_image` table with the related images. Consumers can query for those via the provided query interface.
+
+There does not currently seem to be a reason to expose this over the registry's grpc API, though nothing prevents that in the future.
+
+### Generate and mirror a catalog
+
+`opm` will be extended with a command:
+
+```sh
+$ opm registry build --appregistry-namespace=quay.io/community-operators --to quay.io/ecordell/community-operators-catalog:4.2.1 
+```
+
+`opm registry build` will:
```

Comment:
We probably want to address this if you feel strongly about using `catalog` instead of `registry`:

https://github.com/operator-framework/operator-registry/blob/master/cmd/opm/registry/cmd.go#L11


Also IMO we want to make a distinction between two different layers here. We are planning on having two different commands:

`opm registry` - thing that takes a database file and parameters and outputs a new database file, to be used when you are interacting with the database file itself

`opm index` - thing that takes references to existing index images that contain registry database files, calls `opm registry` as part of a container build, and outputs containers

I do think it makes sense that we use `opm index` in this context (building the disconnected index) instead of `opm registry`, but I think having both is really important to the existing design and implementation of these tools. One outputs database files, one wraps that and also does a container build with the artifact.